(limited) Conditional hail

marius851000 opened the pull request:
**Feature**

This PR addresses the bug/feature described in issue #8379

## Summary
Add:

- The ability to check condition in phrases
- A new syntax for hail (backward compatible). Can have condition, including depending on the hailing ship, and can be weighted (default weight when backward compatibility is used is of 10 per phrase)
- Custom per-ship hail is not modified (thought the conditional phrase can be used)
- Added an extensible way to perform filter query in conditions.
- Add ability to filter on disabled and player-hostile ship

## Usage examples
```
hail "debug hail"
        weight 10000
        message
                word
                        "debug hail: @@@@@@@"

hail "debug hail for hai"
       hailing ship
               government "Hai"
               not disabled
               not hostile

       message
               word
                       "debug hail: @@@@@@@"

hail "debug hail for non-hai"
       to hail
               filter hailingShip
                       not government "Hai"
       message
               word
                       "debug hail: ---------"
```

## Testing Done

A bunch of manual testing on each introduced feature where done. No remaining issues appears. No automated testing implemented.

## Wiki Update

No PR against the Wiki opened yet.

## Performance Impact

Some more time at launch time (in theory. Nothing theoretically worrying)

Some computation when hailing. No performance issue perceived.

A bit more extra RAM needed. Probably not enough that it is a problem. Haven’t profiled memory usage before and after.
I expect 4 or 8 more byte by ConditionSet, in particular.

TomGoodIdea commented:
There are merge conflicts.

Also could you give more details on the new syntax (especially the capabilities of `to hail`)?

bene-dictator commented:
Thanks for working on this, I am really looking forward to seeing this capability in game.

Jurorno9 commented:
I think this could be a really novel idea that could make merchant ships seem more involved in events in the universe. particularity in the FW campaign

>Can have condition, including depending on the hailing ship, and can be weighted (default weight when backward compatibility is used is of 10 per phrase)
What if 2 weights are tied?

JimmyFiryh commented on source/ConditionSet.h:
```diff
@@ -77,6 +89,11 @@ class ConditionSet {
 	// Construct a terminal with a literal value.
 	explicit ConditionSet(int64_t newLiteral);
 
+	// Construct a terminal with a LocationFilter
+	// It is assumed the filter pointer will stay valid for the duration
+	// of this object
```
I would check the Endless Sky [C++ Style Guide](https://endless-sky.github.io/styleguide/styleguide.xml#Comments) for comments.

Brevity should be first and foremost, and multi-line comments are highly discouraged.

marius851000 replied:
I’ll go check my other comments, but this one appear brief enought. It only state the essential info about how to use this function without introducing bugs.

marius851000 replied:
No, I don’t think it have any issue with too long comments.

TomGoodIdea replied:
> and multi-line comments are highly discouraged

```
/* Yes. Multi-line comments are forbidden in our style conventions. */
// Multiple single-line comments, however,
// are not.
```

marius851000 replied:
There is no such /* or */ in my commit outside of the copyright header in my PR as far as I can see. Does that mean there is no problem with my comment?

TomGoodIdea replied:
> There is no such /* or */ in my commit outside of the copyright header in my PR as far as I can see. Does that mean there is no problem with my comment?

yes

tibetiroka commented:
I haven't had the chance to go through these changes properly, but I have a couple concerns:
 - I'm not a fan of adding a ConditionContext to all those functions. Especially not when the only use is to _maybe_  pass a hailing ship to it. Would it be possible to invert the logic? Have the code generating the hails for a ship check the conditions?
 - Hails are per-government. Why are we adding a government condition for them? (Also, don't we have separate hostile and disabled hails in the data? This sounds like overdoing it.)
 - Having a global condition context is bad. It's a global, which is bad, and it's not doing anything useful, which is also bad. It's just a dummy.

marius851000 commented on source/ConditionContext.h:
```diff
@@ -0,0 +1,42 @@
+
+const ConditionContext DEFAULT_CONDITION_CONTEXT = ConditionContext();
+
+/// Information specific for when a ship is hailing the player
+class ConditionContextHailing : public ConditionContext {
+public:
+	ConditionContextHailing(const Ship &hailingShip);
+
+	virtual const Ship * getHailingShip() const override;
+
+private:
+	const Ship &hailingShip;
+};
```
I decided to use multiple class so I get specialized class depending on the context (in case it is wished to extend that method to other context), not wishing to have a long constructor. Thought I could use static function that return a ConditionContext then call the long constructor.

I’m still unsure about performance, but would a few extra memory write would be impactfull? Probably not.

I’ll do that.

Koranir replied:
C++20 has designated initialisers, give each element in the struct a default value and you only override the ones you need to.

marius851000 commented:
> I haven't had the chance to go through these changes properly, but I have a couple concerns:
> 
>     * I'm not a fan of adding a ConditionContext to all those functions. Especially not when the only use is to _maybe_  pass a hailing ship to it. Would it be possible to invert the logic? Have the code generating the hails for a ship check the conditions?
> 
>     * Hails are per-government. Why are we adding a government condition for them? (Also, don't we have separate hostile and disabled hails in the data? This sounds like overdoing it.)
> 
>     * Having a global condition context is bad. It's a global, which is bad, and it's not doing anything useful, which is also bad. It's just a dummy.

The code responsible for hail do check independantly that. The idea of having Phrase have condition was the original implementation, but I kept them once I implemented the hail data, as I thought it might be useful in some cases.

Hail are not per-government. They are declared like News, and if no government filter are put, any government would say them.

It’s a constant global, which is not bad. It can’t be modified by any code (without causing an U.B.), and C++ should be good enough to enforce that in all but the most extreme case where Clang would likely have a warning.

marius851000 commented:
> I think this could be a really novel idea that could make merchant ships seem more involved in events in the universe. particularity in the FW campaign
> 
> > Can have condition, including depending on the hailing ship, and can be weighted (default weight when backward compatibility is used is of 10 per phrase)
> 
> What if 2 weights are tied?

The weight are only statistical weight. If you only have two hail with identifcal weight, you have 1/2 chance of any of the hail being used.

A first hail with a weight of 10 and another one with a weight 30 would have 1/4th chance to use the first hail and 3/4th chance to use the second one.

marius851000 commented:
> There are merge conflicts.
> 
> Also could you give more details on the new syntax (especially the capabilities of `to hail`)?


Yes (I’ll probably reuse that for the wiki):

It change the condition syntax in such a way:

`filter <filteredOn>`: Allow to apply a filter on the element being filtered on. Fail to parse if the element is not recognized, and return 0 (false) if the filtered element is not available in this context.
Valid element for now is limited to "hailingShip". It work like any other LocationFilter on a ship (I can’t find any in the code, actually. (I’ll have to document that too. For now, see LocationFilter.cpp:386)


Phrase can now be disabled if a `to use` token is present at the phrase root and that it return 0/false.


A new structure is added `hail`:

You have to add it to the root of the definition files.
as child, it can have:
 - `weight <number>` impact the chance of this hail being selected between other possible hails. It’s a simple linear weight. The default weight is 10.
 - `hailing ship` is a filter applied on the hailing ship. Kinda a shortcut to `to hail/filter hailingShip`, but can be more easily overridden by plugin. Default to a filter that accept everything.
 - `to hail` is a condition that can apply extra check on whether to use this hail. It act like any other conditions.
 - `message` is a phrase that contain the hail message. The definition is the same as for any other `Phrase` (and, yes, the `to use` also work there, technically speaking)

TomGoodIdea commented:
> `hailing ship` is a filter applied on the hailing ship. Kinda a shortcut to `to hail/filter hailingShip`, but can be more easily overridden by plugin. Default to a filter that accept everything.

Having such "shortcuts" just makes the syntax more complicated to be honest (more keywords to learn, etc.).
Also we use camelCase for variable names, but game data elements should be written in normal words (e.g. `"hailing ship"`, not `hailingShip`).

As a side note: please avoid force-pushing if not necessary.

JimmyFiryh commented:
Inexplicably there is also a significant performance impact experienced. 

CPU: Intel i9-13900H
GPU: NVIDIA RTX 4070 Laptop GPU
RAM: 32GB

Wazn system, v0.10.12 stable sits at 3-4% CPU reported by the game with ships jumping in an out
Same conditions on this PR and it sits at 6-8%, double the CPU usage. 

Hailing ships also results in a 2-3 frame hang. 

These could be issues only I experience, so more testing on this front is probably needed. But, concerning.

marius851000 commented:
So, list of things that I will need to finish (when I have time):

- Check about those multi-line comments
- Verify that the syntax is as good as possible before merging (and thus making this need to be supported long term)
- Double check the performance. Running a loop over the 100 or so possible hail for each of the potentially tens of ship every few seconds is indeed scary.
- Check there is no other comment I forgot to address.

marius851000 commented:
One extra trouble I noticed is that unlike when you define a single phrase for a government that can be reused, if you want to have all the different kind of (let say) Hai use the same hail, it need to enumerate all the different variant of the Hai’s governments. Being able to make government group (with a tag) would solve this, but I’m hesitant to include that in the current PR (it could be added later)

marius851000 commented:
That should be ready for now. I’ve changed a bit the syntax, so that ``"hailing ship"`` is used both directly under ``hail`` and as ``filter "hailing ship"``. I think having both that "shortcut" and that filter is good. One is shorter and easier to override, the other can be used multiple time for complex condition.

I’m still undecided in keeping that option to make some phrase conditional. It was initially used in the initial implementation that worked like that, but is not the case anymore.

I think I’ll remove it.

marius851000 commented:
This should be ready for review. I still have to look at more detail in the performance, but I don’t think it’s an issue. It can run one hail per frame and still run at 60fps (but I’ve been unable to make it run at higher framerate, as if v-sync was always enabled. In any case, any slowdown here should happen only during hail.

warp-core commented:
Updating everything(?) that wants to check conditions is pretty undesirable.
I don't really think it makes sense to test location filters as a condition; if you want the hail to only work based on a location filter, just have the hail check a location filter, don't add support for conditions and then hijack conditions to use a location filter.

marius851000 commented:
> Updating everything(?) that wants to check conditions is pretty undesirable. I don't really think it makes sense to test location filters as a condition; if you want the hail to only work based on a location filter, just have the hail check a location filter, don't add support for conditions and then hijack conditions to use a location filter.

I had thought about it, and I finally decided to keep those condition for allowing more complex logic, such as:
```
or
    filter "hailing ship"
        government "Hai" ...
    and
        filter "hailing ship"
            government "Hai (Unfettered)"
        has "Unfettered Jump Drive 2: offered"
```

 without having to make hacky stuff.

In addition, the name of LocationFilter is misleading. It seems it was originally meant to apply to location, but it can also apply to ship. I apply it on ship, not location (thought they can also check the system and near system). It should probably be renamed to ObjectFilter or something like that. But I think that’s outside of the scope of this PR.

roadrunner56 commented:
Disclaimer that I'm not a code person, but I've been told that this PR contains AI-generated code; is this true?

roadrunner56 commented:
Disclaimer that I'm not a code person, but I've been told that this PR contains AI-generated code; is this true?

marius851000 commented:
I've propably used the same AI-assisted auto completion that I usually use (codium). Rest assured I only used it as a tool to reduce how much code I have to type, and that I only used its code if it was what I would have typed.

(but I've already discussed about that on Discord a while ago, if you're interested)

15 avr. 2025 19:19:54 roadrunner56 ***@***.***>:

> 
>      [Image]*roadrunner56* left a comment (endless-sky/endless-sky#11192)[https://github.com/endless-sky/endless-sky/pull/11192#issuecomment-2806935958]
>     
> 
> Disclaimer that I'm not a code person, but I've been told that this PR contains AI-generated code; is this true?
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/endless-sky/endless-sky/pull/11192#issuecomment-2806935958], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AFMKJZFZIFWVA5BG7JPSHVT2ZU5SXAVCNFSM6AAAAABZGEVMQOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMBWHEZTKOJVHA].
> You are receiving this because you authored the thread.
>

roadrunner56 commented:
Yeah that's fine.

petervdmeer commented:
@marius851000 Thanks for taking up an ambitious project like this.

Proper hails can depend on a lot of complex interactions between NPCs and between NPCs and the player, and it indeed looks silly if some captain is telling a joke when pirates are ambushing their ship. As noted in [my comment on the requesting issue](https://github.com/endless-sky/endless-sky/issues/8379#issuecomment-2889049638), please start with a simpler PR with only a new `hails` rootnode, which has a `message` and a `to show` condition.

The simpler PR will not be a 100% solution for all types of conditional hails, and it would not even solve the silly joking hail from the ambushed captain, but I do feel that this intermediate step will be a good starting point for conditional hails, and does bring us closer to where we need to be.

petervdmeer commented:
I changed this PR to draft and added "Waiting on OP" due the large amount of merge-conflicts, and due to the very large scope the PR currently has. 
Please feel free to continue on this PR (with a smaller scope), or to open a separate PR with just the `hails` rootnode.

marius851000 commented:
> I changed this PR to draft and added "Waiting on OP" due the large amount of merge-conflicts, and due to the very large scope the PR currently has. Please feel free to continue on this PR (with a smaller scope), or to open a separate PR with just the `hails` rootnode.

Is this PR too large? It seems relatively minimal to me while keeping backward compatibility. Now, if backward compatibility is not an issue, I could remove that part. Could also introduce my commits in multiple PR (and do some rebasing), but then it would separate related parts (while having this one PR would make it immediatly usefull).

(and also, I'll do the conflict resolution. With merge instead lf rebase)

marius851000 commented:
I mean, my implementation here is pretty similar to your proposal. Not identical but pretty close.

marius851000 commented:
Arguably, I could move some stuff out of there if judged necessary, like those improved ship filter (which do already exist) and the part that allow to have per-execution specific data in condition sets.
